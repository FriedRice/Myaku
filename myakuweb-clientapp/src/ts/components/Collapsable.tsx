/** @module Wrapper for a collapsable component */

import React from 'react';
import { reflow } from '../utils';

interface CollapsableProps {
    children: React.ReactNode;
    collapsed: boolean;
    onAnimationStart?: () => void;
    onAnimationEnd?: () => void;
}
type Props = CollapsableProps;

interface CollapsableState {
    collapsed: boolean;
    animationStarted: boolean;
}
type State = CollapsableState;

interface CollapsableStyle {
    height?: string;
}


class Collapsable extends React.Component<Props, State> {
    private _collapsableRef: React.RefObject<HTMLDivElement>;

    constructor(props: Props) {
        super(props);
        this.bindEventHandlers();
        this._collapsableRef = React.createRef();

        this.state = {
            collapsed: this.props.collapsed,
            animationStarted: false,
        };
    }

    componentDidUpdate(): void {
        var collapsable = this._collapsableRef.current;
        if (
            this.state.animationStarted
            || collapsable === null
            || this.props.collapsed === this.state.collapsed
        ) {
            return;
        }

        reflow(collapsable);
        this.setState({
            animationStarted: true,
        });
        if (this.props.onAnimationStart) {
            this.props.onAnimationStart();
        }
    }

    bindEventHandlers(): void {
        this.handleAnimationEnd = this.handleAnimationEnd.bind(this);
    }

    handleAnimationEnd(): void {
        if (this._collapsableRef.current !== null) {
            this._collapsableRef.current.removeEventListener(
                'transitionend', this.handleAnimationEnd
            );
        }
        this.setState({
            collapsed: this.props.collapsed,
            animationStarted: false,
        });
        if (this.props.onAnimationEnd) {
            this.props.onAnimationEnd();
        }
    }

    getNoAnimationClassList(): string[] {
        if (this.props.collapsed) {
            return ['collapse'];
        } else {
            return ['collapse', 'show'];
        }
    }

    getAnimationClassList(): string[] {
        if (this.props.collapsed && !this.state.animationStarted) {
            return ['collapse', 'show'];
        } else {
            return ['collapsing'];
        }
    }

    getAnimationInlineStyle(): CollapsableStyle {
        var collapsable = this._collapsableRef.current;
        if (collapsable === null) {
            return {};
        }

        if (!this.props.collapsed) {
            return {
                height: collapsable.scrollHeight + 'px',
            };
        } else if (!this.state.animationStarted) {
            return {
                height: collapsable.getBoundingClientRect()['height'] + 'px',
            };
        } else {
            return {};
        }
    }

    addAnimationEndListener(): void {
        var collapsable = this._collapsableRef.current;
        if (collapsable === null) {
            return;
        }

        collapsable.addEventListener(
            'transitionend', this.handleAnimationEnd
        );
    }

    render(): React.ReactElement {
        var classList: string[] = [];
        var style: CollapsableStyle = {};
        if (this.props.collapsed === this.state.collapsed) {
            classList = this.getNoAnimationClassList();
        } else {
            classList = this.getAnimationClassList();
            style = this.getAnimationInlineStyle();
            if (!this.state.animationStarted) {
                this.addAnimationEndListener();
            }
        }

        return (
            <div
                className={classList.join(' ')}
                style={style}
                ref={this._collapsableRef}
            >
                {this.props.children}
            </div>
        );
    }
}

export default Collapsable;
