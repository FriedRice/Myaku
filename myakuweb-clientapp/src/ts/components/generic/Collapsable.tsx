/**
 * Collapsable component module. See [[Collapsable]].
 */

import React from 'react';
import { reflow } from 'ts/app/utils';

/** Props for the [[Collasable]] component. */
interface CollapsableProps {
    /** Child nodes to make collapsable. */
    children: React.ReactNode;

    /**
     * If true, the component will be collapsed (hiding the children), and
     * if false the component will be uncollapsed (showing the children).
     */
    collapsed: boolean;

    /**
     * If true, changes between collapsed and uncollapsed will be animated, and
     * if false, those changes will be instant with no animation.
     */
    animate: boolean;

    /**
     * Callback function that will be called whenever a collapse or uncollapse
     * animation starts.
     */
    onAnimationStart?: () => void;

    /**
     * Callback function that will be called whenever a collapse or uncollapse
     * animation ends.
     */
    onAnimationEnd?: () => void;
}
type Props = CollapsableProps;

/** State for the [[Collapsable]] component. */
interface CollapsableState {
    /**
     * Whether the collapsable is currently collapsed (true) or uncollapsed
     * (false).
     *
     * @remarks
     * The collapsed prop value indicates the current desired collapsed state,
     * but the collapsed state value indicates the actual current collapsed
     * state.
     * If these two values differ, the component will work to change its
     * collapse/uncollapse state (possibly with an animation depending on the
     * animate prop value) so that the state value will then match the prop
     * value.
     */
    collapsed: boolean;

    /**
     * Whether a collapse/uncollapse animation has started (true) or not
     * (false).
     */
    animationStarted: boolean;
}
type State = CollapsableState;

/** CSS style modified in the process of collapse/uncollapse animation */
interface CollapsableStyle {
    /**
     * The height must be explicitly set inline to trigger the animation at
     * times.
     */
    height?: string;
}


/**
 * Wrapper component to make another component collapsable.
 *
 * @remarks
 * See [[CollapsableProps]] and [[CollapsableState]] for props and state
 * details.
 */
class Collapsable extends React.Component<Props, State> {
    /** Ref for the collapsable container element. */
    private _collapsableRef: React.RefObject<HTMLDivElement>;

    constructor(props: Props) {
        super(props);
        this.bindEventHandlers();
        this._collapsableRef = React.createRef();

        this.state = {
            collapsed: this.props.collapsed,
            animationStarted: false,
        };
    }

    /**
     * Start the collapse/uncollapse animation if necessary.
     *
     * @remarks
     * If the collapsed prop and state values do not currently match and no
     * animation has been started, starts the collapse/uncollapse animation so
     * that the collapsed state value will match the prop value after the
     * animation.
     *
     * If the animate prop is false, just immediately sets the collapsed state
     * value to match the prop value.
     */
    componentDidUpdate(): void {
        var collapsable = this._collapsableRef.current;
        if (
            this.state.animationStarted
            || collapsable === null
            || this.props.collapsed === this.state.collapsed
        ) {
            return;
        }

        if (this.props.animate) {
            reflow(collapsable);
            this.setState({
                animationStarted: true,
            });
            if (this.props.onAnimationStart) {
                this.props.onAnimationStart();
            }
        } else {
            this.setState({
                collapsed: this.props.collapsed,
            });
        }
    }

    /**
     * Bind "this" for the event handlers used by the component.
     */
    bindEventHandlers(): void {
        this.handleAnimationEnd = this.handleAnimationEnd.bind(this);
    }

    /**
     * Callback for updating component state once a collapse/uncollapse
     * animation ends.
     */
    handleAnimationEnd(): void {
        if (this._collapsableRef.current !== null) {
            this._collapsableRef.current.removeEventListener(
                'transitionend', this.handleAnimationEnd
            );
        }
        this.setState({
            collapsed: this.props.collapsed,
            animationStarted: false,
        });
        if (this.props.onAnimationEnd) {
            this.props.onAnimationEnd();
        }
    }

    /**
     * Get the no-animation class list for the collapsable element based on the
     * current props.
     *
     * @returns The list of classes.
     */
    getNoAnimationClassList(): string[] {
        if (this.props.collapsed) {
            return ['collapse'];
        } else {
            return ['collapse', 'show'];
        }
    }

    /**
     * Get the animation class list for the collapsable element based on the
     * current props and state.
     *
     * @returns The list of classes.
     */
    getAnimationClassList(): string[] {
        if (this.props.collapsed && !this.state.animationStarted) {
            return ['collapse', 'show'];
        } else {
            return ['collapsing'];
        }
    }

    /**
     * Get the inline CSS style for the collapsable element based on the
     * current props and state.
     *
     * @returns The inline CSS styles.
     */
    getAnimationInlineStyle(): CollapsableStyle {
        var collapsable = this._collapsableRef.current;
        if (collapsable === null) {
            return {};
        }

        if (!this.props.collapsed) {
            return {
                height: collapsable.scrollHeight + 'px',
            };
        } else if (!this.state.animationStarted) {
            return {
                height: collapsable.getBoundingClientRect()['height'] + 'px',
            };
        } else {
            return {};
        }
    }

    /**
     * Set a callback to be called on a collapse/uncollapse animation end to
     * update component state.
     */
    addAnimationEndListener(): void {
        var collapsable = this._collapsableRef.current;
        if (collapsable === null) {
            return;
        }

        collapsable.addEventListener(
            'transitionend', this.handleAnimationEnd
        );
    }

    render(): React.ReactElement {
        var classList: string[] = [];
        var style: CollapsableStyle = {};
        if (
            this.props.collapsed === this.state.collapsed
            || !this.props.animate
        ) {
            classList = this.getNoAnimationClassList();
        } else {
            classList = this.getAnimationClassList();
            style = this.getAnimationInlineStyle();
            if (!this.state.animationStarted) {
                this.addAnimationEndListener();
            }
        }

        return (
            <div
                className={classList.join(' ')}
                style={style}
                ref={this._collapsableRef}
            >
                {this.props.children}
            </div>
        );
    }
}

export default Collapsable;
