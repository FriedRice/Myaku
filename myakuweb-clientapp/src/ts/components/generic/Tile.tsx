/** @module Content tile */

import React from 'react';

interface TileProps {
    children: React.ReactNode;
    tileClasses?: string;
}

const Tile: React.FC<TileProps> = function(props) {
    var tileClasses = 'tile';
    if (props.tileClasses) {
        tileClasses += ` ${props.tileClasses}`;
    }

    return (
        <section className={tileClasses}>
            {props.children}
        </section>
    );
};

export default Tile;
