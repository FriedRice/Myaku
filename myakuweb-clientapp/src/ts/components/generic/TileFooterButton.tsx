/**
 * TileFooterButton component module. See [[TileFooterButton]].
 */

import React from 'react';

/** Props for the [[TileFooterButton]] component. */
interface TileFooterButtonProps {
    /** Child nodes to show within the tile footer button. */
    children: React.ReactNode;

    /** Callback for when the tile footer button is clicked. */
    onClick: () => void;
}
type Props = TileFooterButtonProps;

/**
 * Component for a footer button for a tile that spans the full width of the
 * tile.
 *
 * @param props - See [[TileFooterButtonProps]].
 */
const TileFooterButton: React.FC<Props> = function(props) {
    return (
        <footer className='tile-footer'>
            <button
                className='button-link tile-footer-button'
                type='button'
                onClick={props.onClick}
            >
                {props.children}
            </button>
        </footer>
    );
};

export default TileFooterButton;
