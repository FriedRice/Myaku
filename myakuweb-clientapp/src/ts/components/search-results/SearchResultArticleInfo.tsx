/**
 * SearchResultArticleInfo component module. See [[SearchResultArticleInfo]].
 */

import { ArticleSearchResult } from 'ts/types/types';
import React from 'react';

import {
    getDaysBetween,
    humanizeDate,
} from 'ts/app/utils';

/** Props for the [[SearchResultArticleInfo]] component. */
interface SearchResultArticleInfoProps {
    searchResult: ArticleSearchResult;
}
type Props = SearchResultArticleInfoProps;


/**
 * Make a time element for the given date.
 *
 * @param date - Date to make a time element for.
 *
 * @returns The time element made for the date.
 */
function makeTimeElement(date: Date): React.ReactElement {
    return (
        <time dateTime={date.toISOString()}>
            {humanizeDate(date)}
        </time>
    );
}

/**
 * Determine if last updated datetime should be displayed for the search
 * result.
 *
 * The last updated datetime is not worth displaying if it's close to the
 * publication date.
 *
 * Additionally, if the publication date is older, the amount of time between
 * publication and the update must be greater in order for the update to be
 * worth displaying.
 *
 * @param searchResult - Search result to determine if its last updated
 * datetime should be displayed.
 *
 * @returns True if the last updated datetime should be displayed, or false if
 * it shouldn't be displayed.
 */
function shouldDisplayLastUpdatedDate(
    searchResult: ArticleSearchResult
): boolean {
    if (searchResult.lastUpdatedDatetime === null) {
        return false;
    }

    var now = new Date();
    var daysSincePublish = getDaysBetween(
        now,
        searchResult.publicationDatetime
    );
    var daysSinceUpdate = getDaysBetween(
        now,
        searchResult.lastUpdatedDatetime
    );
    var daysUpdatedAfter = getDaysBetween(
        searchResult.lastUpdatedDatetime,
        searchResult.publicationDatetime
    );

    if (daysUpdatedAfter < 1 || daysSincePublish <= 7) {
        return false;
    }

    if (
        daysSinceUpdate <= 7
        || daysSincePublish < 180 && daysUpdatedAfter > 30
        || daysSincePublish < 365 && daysUpdatedAfter > 90
        || daysSincePublish < 365 * 2 && daysUpdatedAfter > 180
        || daysUpdatedAfter > 365
    ) {
        return true;
    }

    return false;
}

/**
 * Search result article info list component.
 *
 * @param props - See [[SearchResultArticleInfoProps]].
 */
const SearchResultArticleInfo: React.FC<Props> = function(props) {
    var lastUpdatedDatetimeLi: React.ReactElement | null = null;
    if (shouldDisplayLastUpdatedDate(props.searchResult)) {
        lastUpdatedDatetimeLi = (
            <li key='updated-datetime' className='updated-datetime'>
                {'Updated '}
                {makeTimeElement(
                    props.searchResult.lastUpdatedDatetime as Date
                )}
            </li>
        );
    }

    return (
        <ul className='article-info'>
            <li key='source-name'>
                {props.searchResult.sourceName}
            </li>
            <li key='publication-datetime'>
                {makeTimeElement(props.searchResult.publicationDatetime)}
            </li>
            {lastUpdatedDatetimeLi}
            <li key='instance-count'>
                {props.searchResult.instanceCount.toLocaleString('en-US')}
                {' instances'}
            </li>
        </ul>
    );
};

export default SearchResultArticleInfo;
