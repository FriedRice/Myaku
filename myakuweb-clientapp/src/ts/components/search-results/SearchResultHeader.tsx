/** @module Header component for an article search result */

import { ArticleSearchResult } from 'ts/types/types';
import React from 'react';

interface SearchResultHeaderProps {
    searchResult: ArticleSearchResult;
}
type Props = SearchResultHeaderProps;


const SearchResultHeader: React.FC<Props> = function(props) {
    return (
        <h4 className='main-tile-header' lang='ja'>
            <a href={props.searchResult.sourceUrl}>
                {props.searchResult.title}
            </a>
        </h4>
    );
};

export default SearchResultHeader;
