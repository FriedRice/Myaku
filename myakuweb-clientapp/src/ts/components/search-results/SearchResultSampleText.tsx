/**
 * SearchResultSampleText component module. See [[SearchResultSampleText]].
 */

import { ArticleSampleText } from 'ts/types/types';
import React from 'react';

/** Props for the [[SearchResultSampleText]] component. */
interface SearchResultSampleTextProps {
    /** Sample text content to display in the component */
    sampleText: ArticleSampleText;
}
type Props = SearchResultSampleTextProps;


/**
 * Article search result sample text blockquote component.
 *
 * @param props - See [[SearchResultSampleTextProps]].
 */
const SearchResultSampleText: React.FC<Props> = function(props) {
    var sampleSegments: React.ReactNode[] = [];
    var segments = props.sampleText.segments;
    for (let i = 0; i < segments.length; ++i) {
        if (segments[i].isQueryMatch) {
            sampleSegments.push(
                <span key={`segment-${i}`} className='article-sample-instance'>
                    {segments[i].text}
                </span>
            );
        } else {
            sampleSegments.push(segments[i].text);
        }
    }

    return (
        <blockquote className='article-sample-text'>
            <p lang='ja'>
                {sampleSegments}
            </p>
            <footer>
                {props.sampleText.textStartPos}
            </footer>
        </blockquote>
    );
};

export default SearchResultSampleText;
