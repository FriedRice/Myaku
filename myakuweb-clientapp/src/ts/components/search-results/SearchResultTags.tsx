/** @module Search result tag list component */

import { ArticleSearchResult } from 'ts/types/types';
import React from 'react';

interface SearchResultTagsProps {
    searchResult: ArticleSearchResult;
}
type Props = SearchResultTagsProps;


const SearchResultTags: React.FC<Props> = function(props) {
    var tagLis: React.ReactElement[] = [];
    for (const tag of props.searchResult.tags) {
        tagLis.push(
            <li key={tag} className='tag-badge'>{tag}</li>
        );
    }

    return (
        <ul className='article-tags'>
            {tagLis}
        </ul>
    );
};

export default SearchResultTags;
