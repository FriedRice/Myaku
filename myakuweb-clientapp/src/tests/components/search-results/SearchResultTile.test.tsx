/**
 * Tests for the SearchResultTile component.
 * @module tests/components/search-results/SearchResultTile.test
 */

import Collapsable from 'ts/components/generic/Collapsable';
import React from 'react';
import SearchResultArticleInfo from
    'ts/components/search-results/SearchResultArticleInfo';
import SearchResultHeader from
    'ts/components/search-results/SearchResultHeader';
import SearchResultSampleText from
    'ts/components/search-results/SearchResultSampleText';
import SearchResultTags from 'ts/components/search-results/SearchResultTags';
import SearchResultTile from 'ts/components/search-results/SearchResultTile';
import Tile from 'ts/components/generic/Tile';
import TileFooterButton from 'ts/components/generic/TileFooterButton';
import { act } from 'react-dom/test-utils';
import { expectComponent } from 'tests/testUtils';
import { getSearchResultPageDataClone } from 'tests/testData';

import {
    ShallowWrapper,
    shallow,
} from 'enzyme';

const testResultPage = getSearchResultPageDataClone();
var searchQuery = testResultPage.search.query;
var searchResultWithMoreSampleText = testResultPage.results[0];
var searchResultWithoutMoreSampleText = testResultPage.results[1];
beforeEach(function() {
    const testResultPage = getSearchResultPageDataClone();
    searchQuery = testResultPage.search.query;
    searchResultWithMoreSampleText = testResultPage.results[0];
    searchResultWithoutMoreSampleText = testResultPage.results[1];
});

function getTile(wrapper: ShallowWrapper): ShallowWrapper {
    const tile = wrapper.find(Tile);
    expect(tile).toHaveLength(1);
    return tile;
}

function expectMoreSampleTextCollapsed(
    wrapper: ShallowWrapper, collapsed: boolean,
    animate: boolean = expect.any(Boolean)
): void {
    expectComponent(getTile(wrapper), Collapsable, {
        collapsed: collapsed,
        animate: animate,
        onAnimationEnd: expect.any(Function),
        children: expect.anything(),
    });
}

function toggleMoreSampleTextCollapse(wrapper: ShallowWrapper): void {
    const collapseToggle = getTile(wrapper).find(TileFooterButton);
    expect(collapseToggle).toHaveLength(1);
    act(() => collapseToggle.props().onClick());
}

function mockCollapseOnAnimationEndEvent(wrapper: ShallowWrapper): void {
    const collapsable = getTile(wrapper).find(Collapsable);
    expect(collapsable).toHaveLength(1);

    const onAnimationEnd = collapsable.props().onAnimationEnd;
    if (onAnimationEnd === undefined) {
        throw new Error(
            'Search options collapsable has no onAnimationEnd callback set'
        );
    }
    act(() => onAnimationEnd());
}

function expectCollapseToggleButtonText(
    wrapper: ShallowWrapper, collapsed: boolean
): void {
    var startText = '';
    if (collapsed) {
        startText = 'Show more ';
    } else {
        startText = 'Show less ';
    }

    const children = getTile(wrapper).find(TileFooterButton).children();
    expect(children).toHaveLength(3);
    expect(children.at(0).text()).toBe(startText);
    expect(children.at(1).name()).toBe('span');
    expect(children.at(1).props().lang).toBe('ja');
    expect(children.at(1).text()).toBe(searchQuery);
    expect(children.at(2).text()).toBe(' instances from this article');
}


describe('<SearchResultTile /> render', function() {
    var moreSampleTextTile = shallow(<div />);
    var noMoreSampleTextTile = shallow(<div />);
    beforeEach(function() {
        const wrapperWithMoreSampleText = shallow(
            <SearchResultTile
                searchQuery={searchQuery}
                searchResult={searchResultWithMoreSampleText}
            />
        );
        moreSampleTextTile = wrapperWithMoreSampleText.find(Tile);

        const wrapperWithoutMoreSampleText = shallow(
            <SearchResultTile
                searchQuery={searchQuery}
                searchResult={searchResultWithoutMoreSampleText}
            />
        );
        noMoreSampleTextTile = wrapperWithoutMoreSampleText.find(Tile);
    });

    it('renders loading tile if null search result', function() {
        const wrapper = shallow(
            <SearchResultTile
                searchQuery={null}
                searchResult={null}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders header if non-null search result', function() {
        expectComponent(moreSampleTextTile, SearchResultHeader, {
            searchResult: searchResultWithMoreSampleText,
        });
        expectComponent(noMoreSampleTextTile, SearchResultHeader, {
            searchResult: searchResultWithoutMoreSampleText,
        });
    });

    it('renders article info if non-null search result', function() {
        expectComponent(moreSampleTextTile, SearchResultArticleInfo, {
            searchResult: searchResultWithMoreSampleText,
        });
        expectComponent(noMoreSampleTextTile, SearchResultArticleInfo, {
            searchResult: searchResultWithoutMoreSampleText,
        });
    });

    it('renders tags if non-null search result', function() {
        expectComponent(moreSampleTextTile, SearchResultTags, {
            searchResult: searchResultWithMoreSampleText,
        });
        expectComponent(noMoreSampleTextTile, SearchResultTags, {
            searchResult: searchResultWithoutMoreSampleText,
        });
    });

    it('renders main sample text if non-null search result', function() {
        var mainSampleText = moreSampleTextTile.find(
            SearchResultSampleText
        ).first();
        expect(mainSampleText).toHaveLength(1);
        expect(mainSampleText.props()).toStrictEqual({
            sampleText: searchResultWithMoreSampleText.mainSampleText,
        });

        mainSampleText = noMoreSampleTextTile.find(
            SearchResultSampleText
        ).first();
        expect(mainSampleText).toHaveLength(1);
        expect(mainSampleText.props()).toStrictEqual({
            sampleText: searchResultWithoutMoreSampleText.mainSampleText,
        });
    });

    it('renders all more sample text if given', function() {
        const searchResult = searchResultWithMoreSampleText;
        const sampleTexts = moreSampleTextTile.find(SearchResultSampleText);

        // Add 1 to account for the main sample text on the tile.
        expect(sampleTexts).toHaveLength(
            searchResult.moreSampleTexts.length + 1
        );

        // Start i from 1 to skip the main sample text on the tile.
        for (let i = 1; i < sampleTexts.length; ++i) {
            expect(sampleTexts.at(i).props()).toStrictEqual({
                sampleText: searchResult.moreSampleTexts[i - 1],
            });
        }
    });

    it(
        'does not render collapsable if only main sample text given',
        function() {
            const collapsable = noMoreSampleTextTile.find(Collapsable);
            expect(collapsable).toHaveLength(0);
        }
    );

    it('renders toggle if more sample text given', function() {
        expectComponent(moreSampleTextTile, TileFooterButton, {
            onClick: expect.any(Function),
            children: expect.anything(),
        });
    });

    it('does not render toggle if more sample text not given', function() {
        const tileFooterButton = noMoreSampleTextTile.find(TileFooterButton);
        expect(tileFooterButton).toHaveLength(0);
    });
});

describe('<SearchResultTile /> more sample text collapsable', function() {
    var wrapper = shallow(<div />);
    beforeEach(function() {
        wrapper = shallow(
            <SearchResultTile
                searchQuery={searchQuery}
                searchResult={searchResultWithMoreSampleText}
            />
        );
    });

    it('collapses more sample text by default on mount', function() {
        expectMoreSampleTextCollapsed(wrapper, true);
    });

    it('handles more sample text collapse toggle', function() {
        expectMoreSampleTextCollapsed(wrapper, true);

        toggleMoreSampleTextCollapse(wrapper);
        expectMoreSampleTextCollapsed(wrapper, false, true);

        mockCollapseOnAnimationEndEvent(wrapper);
        toggleMoreSampleTextCollapse(wrapper);
        expectMoreSampleTextCollapsed(wrapper, true, true);

        mockCollapseOnAnimationEndEvent(wrapper);
        toggleMoreSampleTextCollapse(wrapper);
        expectMoreSampleTextCollapsed(wrapper, false, true);
    });

    it('ignores more sample text collapse toggle if animating', function() {
        expectMoreSampleTextCollapsed(wrapper, true);

        toggleMoreSampleTextCollapse(wrapper);
        expectMoreSampleTextCollapsed(wrapper, false, true);

        toggleMoreSampleTextCollapse(wrapper);
        expectMoreSampleTextCollapsed(wrapper, false, true);

        mockCollapseOnAnimationEndEvent(wrapper);
        toggleMoreSampleTextCollapse(wrapper);
        expectMoreSampleTextCollapsed(wrapper, true, true);

        toggleMoreSampleTextCollapse(wrapper);
        expectMoreSampleTextCollapsed(wrapper, true, true);
    });

    it('sets footer button text to match collapsed state', function() {
        expectCollapseToggleButtonText(wrapper, true);

        toggleMoreSampleTextCollapse(wrapper);
        expectCollapseToggleButtonText(wrapper, false);

        mockCollapseOnAnimationEndEvent(wrapper);
        toggleMoreSampleTextCollapse(wrapper);
        expectCollapseToggleButtonText(wrapper, true);

        mockCollapseOnAnimationEndEvent(wrapper);
        toggleMoreSampleTextCollapse(wrapper);
        expectCollapseToggleButtonText(wrapper, false);
    });
});
