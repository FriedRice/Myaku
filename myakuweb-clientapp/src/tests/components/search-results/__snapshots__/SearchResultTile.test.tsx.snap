// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultTile /> render renders loading tile if null search result 1`] = `
<Tile
  loadingHeight="13rem"
  tileClasses="result-tile"
/>
`;
